[SPARK-5778] throw if nonexistent metrics config file provided

ryan-williams opened the pull request:
previous behavior was to log an error; this is fine in the general
case where no `spark.metrics.conf` parameter was specified, in which
case a default `metrics.properties` is looked for, and the execption
logged and suppressed if it doesn't exist.

if the user has purposefully specified a metrics.conf file, however,
it makes more sense to show them an error when said file doesn't
exist.

ryan-williams committed:
throw if nonexistent metrics config file provided

previous behavior was to log an error; this is fine in the general
case where no `spark.metrics.conf` parameter was specified, in which
case a default `metrics.properties` is looked for, and the execption
logged and suppressed if it doesn't exist.

if the user has purposefully specified a metrics.conf file, however,
it makes more sense to show them an error when said file doesn't
exist.

SparkQA commented:
[Test build #27368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27368/consoleFull) for   PR 4571 at commit [`105288e`](https://github.com/apache/spark/commit/105288e2c661ed631e64cfdc3aeb79b2aea858eb).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27368/consoleFull) for   PR 4571 at commit [`105288e`](https://github.com/apache/spark/commit/105288e2c661ed631e64cfdc3aeb79b2aea858eb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27368/
Test FAILed.

rxin commented on core/src/main/scala/org/apache/spark/metrics/MetricsConfig.scala:
```diff
@@ -47,21 +47,27 @@ private[spark] class MetricsConfig(val configFile: Option[String]) extends Loggi
-      is = configFile match {
-        case Some(f) => new FileInputStream(f)
-        case None => Utils.getSparkClassLoader.getResourceAsStream(METRICS_CONF)
+    (configFile match {
```
I've been thinking about style quite a bit recently as I'm working on a style guide. In this case, you'd want to break the chain down with intermediate variables. Otherwise the cognitive complexity of the chaining/nesting makes it hard to understand.

I wrote 3 versions of them, with each version gradually becoming more Java-like. I think either would be fine here.

version 1
```scala
val isOpt: Option[InputStream] = configFile.map(new FileInputStream(_)).orElse {
  try {
    Option(Utils.getSparkClassLoader.getResourceAsStream(METRICS_CONF))
  } catch {
    case e: Exception =>
      logError("Error loading default configuration file", e)
      None
  }
}

isOpt.foreach { is =>
  try {
    properties.load(is)
  } finally {
    is.close()
  }
}
```

version 2
```scala
var isOpt: Option[InputStream] = configFile.map { f =>
  logInfo("Loading MetricsConfig file: $f")
  new FileInputStream(f)
}

if (!isOpt.isDefined) {
  try {
    is = Option(Utils.getSparkClassLoader.getResourceAsStream(METRICS_CONF))
  } catch {
    case e: Exception =>
      logError("Error loading default configuration file", e)
  }
}

isOpt.foreach { is =>
  try {
    properties.load(is)
  } finally {
    is.close()
  }
}
```

version 3
```scala
var is: InputStream = configFile.map(new FileInputStream(_)).orNull

if (is == null) {
  try {
    is = Option(Utils.getSparkClassLoader.getResourceAsStream(METRICS_CONF))
  } catch {
    case e: Exception =>
      logError("Error loading default configuration file", e)
  }
}

if (is != null) {
  try {
    properties.load(is)
  } finally {
    is.close()
  }
}
```

ryan-williams replied:
cool, thanks, went with the first one.

incidentally I had a `Some()` where I should have had an `Option()`, which your candidates all fix, that I think was breaking the tests

ryan-williams committed:
add "metrics.properties" to .rat-excludes

it's common to have this file laying around as the default config file
for MetricsSystem, but RAT doesn't like its lack of an Apache header.

ryan-williams committed:
metrics code review feedback

ryan-williams committed:
throw if nonexistent Sink class is specified

SparkQA commented:
[Test build #27376 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27376/consoleFull) for   PR 4571 at commit [`b98e325`](https://github.com/apache/spark/commit/b98e3257f04449aee35175499f2abbf3e66a4a03).
 * This patch merges cleanly.

ryan-williams commented:
I threw in a couple of other trivial things:
* `throw` if a bad `Sink` class is specified
* exclude `conf/metrics.properties`, if present, from RAT checks

lmk if you want me to make those separate JIRAs/PRs

SparkQA commented:
[Test build #27378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27378/consoleFull) for   PR 4571 at commit [`45f1d3e`](https://github.com/apache/spark/commit/45f1d3ef1f1363f8e07a69d9ba52b08a9dfcec82).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27376 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27376/consoleFull) for   PR 4571 at commit [`b98e325`](https://github.com/apache/spark/commit/b98e3257f04449aee35175499f2abbf3e66a4a03).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27376/
Test FAILed.

SparkQA commented:
[Test build #27378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27378/consoleFull) for   PR 4571 at commit [`45f1d3e`](https://github.com/apache/spark/commit/45f1d3ef1f1363f8e07a69d9ba52b08a9dfcec82).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logError("Sink class " + classPath + " cannot be instantialized")`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27378/
Test FAILed.

ryan-williams committed:
fix tests

SparkQA commented:
[Test build #27399 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27399/consoleFull) for   PR 4571 at commit [`1b74f69`](https://github.com/apache/spark/commit/1b74f69d08854af7870bf169260e0700ee4016ab).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27399 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27399/consoleFull) for   PR 4571 at commit [`1b74f69`](https://github.com/apache/spark/commit/1b74f69d08854af7870bf169260e0700ee4016ab).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logError("Sink class " + classPath + " cannot be instantialized")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27399/
Test PASSed.

rxin commented:
This looks good. @pwendell I'm not sure which branch you want this in, so I will let you do the merge.

ryan-williams committed:
rename METRICS_CONF: DEFAULT_METRICS_CONF_FILENAME

ryan-williams committed:
private-ize some MetricsConfig members

ryan-williams commented:
I just rebased and tacked a couple of trivial commits to `MetricsConfig` onto the end of this, jfyi

rxin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #27464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27464/consoleFull) for   PR 4571 at commit [`5bccb14`](https://github.com/apache/spark/commit/5bccb142b85411b6c76f2faa0215d1f2dc85374e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27464/consoleFull) for   PR 4571 at commit [`5bccb14`](https://github.com/apache/spark/commit/5bccb142b85411b6c76f2faa0215d1f2dc85374e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logError("Sink class " + classPath + " cannot be instantialized")`
  * `  case class Params(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27464/
Test FAILed.

ryan-williams commented:
looks like similar sorts of probably-spurious to those that I saw on #4573; thoughts, @rxin? re-test?

rxin commented:
Jenkins, retest this please

SparkQA commented:
[Test build #27468 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27468/consoleFull) for   PR 4571 at commit [`5bccb14`](https://github.com/apache/spark/commit/5bccb142b85411b6c76f2faa0215d1f2dc85374e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27468 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27468/consoleFull) for   PR 4571 at commit [`5bccb14`](https://github.com/apache/spark/commit/5bccb142b85411b6c76f2faa0215d1f2dc85374e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logError("Sink class " + classPath + " cannot be instantialized")`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27468/
Test FAILed.

ryan-williams commented:
seems like another spurious failure to me. That one passed in the previous build

pwendell commented:
Jenkins, retest this please. (sorry @ryan-williams we're trying to get tests in better shape).

SparkQA commented:
[Test build #27529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27529/consoleFull) for   PR 4571 at commit [`5bccb14`](https://github.com/apache/spark/commit/5bccb142b85411b6c76f2faa0215d1f2dc85374e).
 * This patch merges cleanly.

pwendell commented:
Overall, looks good to me. I do like the idea of failing here, since otherwise it can be very confusing to people why their settings are not being picked up. To avoid surprise here I think it's best to avoid backporting this patch (once merged) into older versions and just enforce it in new versions.

One other issue, it's not super well documented in the existing code base what type of error propagation we expect when initializing a spark environment on the executors. It would be good to verify that this gets thrown up the chain in a way that gets presented nicely to the user. I wouldn't block the patch on that (it's not newly introduced here), but having nicer internal documentation about the expectations that callers should have if SparkEnv's fail to initialize would be nice.

SparkQA commented:
[Test build #27529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27529/consoleFull) for   PR 4571 at commit [`5bccb14`](https://github.com/apache/spark/commit/5bccb142b85411b6c76f2faa0215d1f2dc85374e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logError("Sink class " + classPath + " cannot be instantialized")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27529/
Test PASSed.

hammer commented:
@pwendell @rxin looks ready to merge?

pwendell commented:
Pulling it in now, thanks ryan!
